Auto-respond to everyone pings.

MrAwesomeRocks opened the pull request:
# Everyone Ping Auto-Responder
### Closes #1131 

The result of sending an `@everyone` ping would look like this: 
![result of ping in server](https://user-images.githubusercontent.com/42477863/91647869-ca7b7700-ea25-11ea-94cb-2878835ef812.png)
The user would also get DMed the default message for an anti-spam rule:
![result of ping in DM](https://user-images.githubusercontent.com/42477863/91647910-7ae97b00-ea26-11ea-917f-cc4c2e8fb2f5.png)

Also, this filter detects ***all*** cases of `@everyone` pings.

## Notes:
Set the `max` value in the config file to zero for this to work.

The server member count updates automatically.

lemonsaurus commented on bot/rules/everyone_ping.py:
```diff
@@ -0,0 +1,37 @@
+    """Detects if a user has sent an '@everyone' ping."""
+    relevant_messages = tuple(msg for msg in recent_messages if msg.author == last_message.author)
+
+    ev_msgs_ct = 0
```
Please don't use cryptic, abbreviated variable names like this.

MrAwesomeRocks commented:
Changed the message to show the full member count, not just thousands: 
![ping response](https://user-images.githubusercontent.com/42477863/91659189-0d743380-ea94-11ea-8bfc-678145fa0ab6.png)

### Notes:
- This value will be the actual member count.
- The number will also have a comma every three digits.

lemonsaurus commented on bot/rules/everyone_ping.py:
```diff
@@ -0,0 +1,37 @@
+        # Send the user an embed giving them more info:
+        embed_text = textwrap.dedent(
+            f"""
+            Please don't try to ping {last_message.guild.member_count:,} people.
+            **It will not have good results.**
```
I don't like the `It will not have good results` part of this, that feels like we're saying it's okay to ping 95000 people but it won't get you the results you were hoping for.

Instead of this second line, maybe we can give this embed a headline. For example, you could use a random headline from `bot.constants.NEGATIVE_REPLIES`?

lemonsaurus replied:
With the second line removed, this can just be a regular f string. There's no need for textwrap dedent.

MrAwesomeRocks replied:
By `It will not have good results`, I was trying to say that people would get upset if you pinged them.

I'll look at `bot.constants.NEGATIVE_REPLIES`, didn't know that existed.

lemonsaurus commented on bot/rules/everyone_ping.py:
```diff
@@ -0,0 +1,37 @@
+        """
+        )
+        embed = Embed(description=embed_text, colour=Colours.soft_red)
+        await last_message.channel.send(f"Hey {last_message.author.mention}!", embed=embed)
```
Feels weird to me that we're pinging the user before the embed. It makes the embed sandwiched between two plaintext lines, and also makes us ping the user twice instead of just once. I don't think we need it.
```suggestion
        await last_message.channel.send(embed=embed)
```

lemonsaurus reviewed (changes requested):
I have a few minor suggestions here.

MrAwesomeRocks commented:
The bot now responds like this:
![image](https://user-images.githubusercontent.com/42477863/91660092-ad808b80-ea99-11ea-8d68-c620af2ec9b0.png)

ks129 commented on bot/rules/everyone_ping.py:
```diff
@@ -0,0 +1,42 @@
+        # Send the user an embed giving them more info:
+        embed_text = textwrap.dedent(
+            f"""
+            **{random.choice(NEGATIVE_REPLIES)}**
```
I'd suggest to make this as embed title

MrAwesomeRocks commented:
`NEGATIVE_REPLIES` now moved to title of embed instead of bold.
![ping reply embed](https://user-images.githubusercontent.com/42477863/91662220-9f853780-eaa6-11ea-9fe4-e773280f5885.png)

lemonsaurus commented:
@MrAwesomeRocks 
Remember to resolve conversations in this thread when you've addressed them.

MrAwesomeRocks commented:
@lemonsaurus 
Didn't know about that. All the conversations have been resolved.

lemonsaurus reviewed (changes requested):
You've implemented this as a frequency filter, which I find a bit weird. It's not really relevant for our community to be able to configure this filter to allow only 2 `@everyone` per 10 seconds, for example. I can't see why anyone would ever need that.

Perhaps we can just simplify this pull request by removing the frequency aspect completely from this filter, and removing the corresponding configuration option, too.

MrAwesomeRocks commented:
@lemonsaurus 
The issue is, to implement this as an anti-spam rule, in anitspam.py, there is this code that checks for those two config options: 
```python
for required_key in ('interval', 'max'):
    if required_key not in config:
        log.error(
            f"`{required_key}` is required but was not "
                f"set in rule `{name}`'s configuration."
            )
            validation_errors[name] = f"Key `{required_key}` is required but not set for rule `{name}`"
```
And the `interval` rule is used to determine how often to check messages: 
```python
# Fetch the rule configuration with the highest rule interval.
max_interval_config = max(
    AntiSpamConfig.rules.values(),
    key=itemgetter('interval')
)
max_interval = max_interval_config['interval']
```
Which is why I have this implemented as a frequency filter.

kwzrd commented on bot/rules/everyone_ping.py:
```diff
@@ -0,0 +1,36 @@
+        # Make embed:
+        embed = Embed(title=random.choice(NEGATIVE_REPLIES), description=embed_text, colour=Colours.soft_red)
+
+        # Send embed:
+        await last_message.channel.send(embed=embed)
```
I do not believe this comment is necessary, it just repeats what the code says. Similar above with "Make embed".

kwzrd reviewed (commented):
The current solution will punish me if I send a message containing `@everyone` as a markdown codeblock, which is unfortunate because there is clearly no intention to ping, e.g. when discussing this feature.

I also believe this will not react to pings constructed as `<@&{guild_id}>`.

Edit: It looks like Discord may not be allowing pinging the everyone role with this syntax, although it almost looks a bit like a bug? Either way, if someone does send that string, I'd argue that there is a clear intention to ping.

lemonsaurus commented:
okay, after looking at antispam.py a bit, I guess refactoring to allow non-frequency spam filters seems like it was would be out of scope.

I guess we will go with this solution.

MrAwesomeRocks commented:
- Switched to regex to avoid punishing users for codeblocks.
- Also now catches `<@&{guild_id}>` style pings.
- Deleted the redundant comments.
- Changed the message to be less like a frequency anti-spam rule: 
![everyone ping message reply](https://user-images.githubusercontent.com/42477863/91678791-16a0e700-eb0c-11ea-9f2e-cf0b0746af8d.png)

MrAwesomeRocks commented:
My regex was catching some incorrect strings, such as ```markdown

MrAwesomeRocks commented:
My regex was catching some incorrect strings, such as 
```markdown
```@everyone
here is some text
here is more text```
``` 
Now the regex doesn't catch those incorrect strings.

Aniperk commented on bot/cogs/antispam.py:
```diff
@@ -34,7 +34,8 @@
     'newlines': rules.apply_newlines,
-    'role_mentions': rules.apply_role_mentions
+    'role_mentions': rules.apply_role_mentions,
+    'everyone_ping': rules.apply_everyone_ping,
```
line 38: at the end im pretty sure there doesn't need to be a comma, probably just nit picking here :)

kwzrd replied:
This diff is a good example of how the trailing comma is useful. Because line `37` did not have the trailing comma, it had to be changed in order to add another element.

lemonsaurus replied:
Yep, leave this trailing comma in - in fact, it's encouraged by PEP8 as well.

Aniperk replied:
Alright, i was just thinking it might not need to be there, im just someone who works in environments where using as least space as possible is the goal. :)

MrAwesomeRocks commented:
All of kwzrd's concerns about punishing users who put `@everyone` in codeblocks should be fixed now.

MrAwesomeRocks commented:
Wait I need to fix some regex.

MrAwesomeRocks commented:
Now it catches `@everyone` with text next to it, such as `hello@everyone`.

lemonsaurus commented:
@MrAwesomeRocks 

How will this affect staff members? They should be immune to it. Can you confirm that they will be? I think they're immune to all anti-spam, but I'd just like a confirmation.

LordHemlock reviewed (approved):
Seems to work as intended in testing.  Giving it the green light.

lemonsaurus reviewed (approved):
Alright, seems fine now. let's get this merged.
